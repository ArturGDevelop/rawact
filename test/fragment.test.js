import React from "react";
import ReactDOM from "react-dom";
import renderHelper from "./renderHelper";

const Hello = ({ name }) => <React.Fragment>Hello {name}!</React.Fragment>;
const GoodBye = ({ name }) => (
	<React.Fragment>
		Good<b>B</b>ye {name}!
	</React.Fragment>
);

const ChangingChildren = ({ type, name }) => {
	const children = [
		type === "hello" ? <Hello name={name} /> : <GoodBye name={name} />
	];
	if (type === "goodbye") {
		children.push("(and sorry)");
	}
	return <React.Fragment>{children}</React.Fragment>;
};

const ChangingChildrenElement = ({ isDiv }) => {
	return (
		<React.Fragment>
			Testing...
			{isDiv ? <div /> : <span />}
		</React.Fragment>
	);
};

describe("fragments", () => {
	describe("rendering at the top", () => {
		it("renders correctly", () => {
			const { element } = renderHelper(<Hello name="world" />);

			expect(element).toMatchSnapshot();
		});

		it("updates correctly", () => {
			const { element } = renderHelper(<Hello name="world" />);

			ReactDOM.render(<Hello name="me" />, element);

			expect(element).toMatchSnapshot();
		});

		it("updates correctly, changing a child", () => {
			const { element } = renderHelper(
				<ChangingChildrenElement isDiv={false} />
			);

			ReactDOM.render(<ChangingChildrenElement isDiv />, element);

			expect(element).toMatchSnapshot();
		});

		// TODO - null reference exception in replaceNode
		xit("adds new children", () => {
			const { element } = renderHelper(
				<ChangingChildren name="world" type="hello" />
			);

			ReactDOM.render(<ChangingChildren name="me" type="goodbye" />, element);

			expect(element).toMatchSnapshot();
		});

		// TODO - null reference exception in replaceNode
		xit("removes children", () => {
			const { element } = renderHelper(
				<ChangingChildren name="world" type="goodbye" />
			);

			ReactDOM.render(<ChangingChildren name="me" type="hello" />, element);

			expect(element).toMatchSnapshot();
		});

		it("removes non react children", () => {
			const element = document.createElement("div");
			element.innerHTML = "This<span>Rogue</span>Test";
			ReactDOM.render(<Hello name="world" />, element);

			expect(element).toMatchSnapshot();
		});
	});

	describe("rendering under an element", () => {
		it("renders correctly", () => {
			const { element } = renderHelper(
				<div>
					<Hello name="world" />
				</div>
			);

			expect(element).toMatchSnapshot();
		});

		it("updates correctly", () => {
			const { element } = renderHelper(
				<div>
					<Hello name="world" />
				</div>
			);

			ReactDOM.render(
				<div>
					<Hello name="me" />
				</div>,
				element
			);

			expect(element).toMatchSnapshot();
		});

		it("updates correctly, changing a child", () => {
			const { element } = renderHelper(
				<div>
					<ChangingChildrenElement isDiv={false} />
				</div>
			);

			ReactDOM.render(
				<div>
					<ChangingChildrenElement isDiv />
				</div>,
				element
			);

			expect(element).toMatchSnapshot();
		});

		it("adds new children", () => {
			const { element } = renderHelper(
				<div>
					<ChangingChildren name="world" type="hello" />
				</div>
			);

			ReactDOM.render(
				<div>
					<ChangingChildren name="me" type="goodbye" />
				</div>,
				element
			);

			expect(element).toMatchSnapshot();
		});

		it("removes children", () => {
			const { element } = renderHelper(
				<div>
					<ChangingChildren name="world" type="goodbye" />
				</div>
			);

			ReactDOM.render(
				<div>
					<ChangingChildren name="me" type="hello" />
				</div>,
				element
			);

			expect(element).toMatchSnapshot();
		});

		it("moves up", () => {
			const { element } = renderHelper(
				<div>
					<Hello name="world" />
				</div>
			);

			ReactDOM.render(<Hello name="world" />, element);

			expect(element).toMatchSnapshot();
		});

		it("moves down", () => {
			const { element } = renderHelper(<Hello name="world" />);

			ReactDOM.render(
				<div>
					<Hello name="world" />
				</div>,
				element
			);

			expect(element).toMatchSnapshot();
		});
	});
});
