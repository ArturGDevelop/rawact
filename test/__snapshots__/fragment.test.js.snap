// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fragments rendering at the top adds new children 1`] = `
<div>
  Good
  <b>
    B
  </b>
  ye 
  me
  !
  (and sorry)
</div>
`;

exports[`fragments rendering at the top removes children 1`] = `
<div>
  Hello 
  me
  !
</div>
`;

exports[`fragments rendering at the top removes non react children 1`] = `
<div>
  This
  <span>
    Rogue
  </span>
  Test
</div>
`;

exports[`fragments rendering at the top renders correctly 1`] = `
<div>
  Hello 
  world
  !
</div>
`;

exports[`fragments rendering at the top updates correctly 1`] = `
<div>
  Hello 
  world
  !
</div>
`;

exports[`fragments rendering at the top updates correctly, changing a child 1`] = `
<div>
  Testing...
  <span />
</div>
`;

exports[`fragments rendering under an element adds new children 1`] = `
<div>
  <div>
    Hello 
    world
    !
  </div>
</div>
`;

exports[`fragments rendering under an element moves down 1`] = `
<div>
  Hello 
  world
  !
</div>
`;

exports[`fragments rendering under an element moves up 1`] = `
<div>
  <div>
    Hello 
    world
    !
  </div>
</div>
`;

exports[`fragments rendering under an element removes children 1`] = `
<div>
  <div>
    Good
    <b>
      B
    </b>
    ye 
    world
    !
    (and sorry)
  </div>
</div>
`;

exports[`fragments rendering under an element renders correctly 1`] = `
<div>
  <div>
    Hello 
    world
    !
  </div>
</div>
`;

exports[`fragments rendering under an element updates correctly 1`] = `
<div>
  <div>
    Hello 
    world
    !
  </div>
</div>
`;

exports[`fragments rendering under an element updates correctly, changing a child 1`] = `
<div>
  <div>
    Testing...
    <span />
  </div>
</div>
`;
